### ISSUE TITLE
Help/Advice: malloc: Heap corruption detected when using Open SSL with CivetWeb Websockets

### ISSUE THREAD

**Issue by jackfruit2** [civetweb/civetweb#1050]
Hi,

I'm playing around with Civetweb at the moment, running secure websocket connections in conjunction with regular Https and I have stumbled across an issue, to which I'm stumped.

I'm running some heavy traffic via the websocket and then I make another connection whether its a straight HTTPS or WSS connection and it crashes with Heap corruption detected, intermittently. When I run the same test using HTTP and WS connections the server holds up just fine. Also I'm using Apache JMeter to soak test the civetweb server.

I'm wondering if anyone could make any suggestions/pointers (no pun) as to where I should focus my attention when investigating this issue. I suspect it has something to do with either the connection being dropped or socket death, where civetweb is assuming its active (pure conjecture). 

FYI  I'm using a MacBook OSX and Open SSL 1.1.1k. 

Below I have captured some of the logging for inspection:

```
*** 1640124996.156991000 123145466884096 accept_new_connection:19688: Accepted socket 12
*** 1640124996.157055000 123145466884096 produce_socket:19397: queued socket 12
*** 1640124996.157082000 123145466884096 accept_new_connection:19688: Accepted socket 14
*** 1640124996.157098000 123145466884096 produce_socket:19397: queued socket 14
*** 1640124996.157254000 123145464475648 consume_socket:19353: grabbed socket 12, going busy
*** 1640124996.157307000 123145464475648 worker_thread_run:19511: Incomming SSL connection from 192.168.10.38
*** 1640124996.157334000 123145464705024 consume_socket:19353: grabbed socket 14, going busy
TestWebServer(37742,0x700009aa7000) malloc: Heap corruption detected, free list is damaged at 0x6000002eea40
*** Incorrect guard value: 25190271937024
*** 1640124996.157343000 123145466884096 accept_new_connection:19688: Accepted socket 15
*** 1640124996.157380000 123145466884096 produce_socket:19397: queued socket 15
*** 1640124996.157434000 123145466884096 accept_new_connection:19688: Accepted socket 17TestWebServer(37742,0x700009aa7000) malloc: *** set a breakpoint in malloc_error_break to debug

*** 1640124996.157459000 123145466884096 produce_socket:19397: queued socket 17
*** 1640124996.157482000 123145466884096 accept_new_connection:19688: Accepted socket 18
*** 1640124996.157497000 123145466884096 produce_socket:19397: TestWebServer(37742,0x700009aa7000) malloc: Heap corruption detected, free list is damaged at 0x6000002eea40
*** Incorrect guard value: 25190271937024 
```

<img width="1002" alt="Screenshot 2021-12-21 at 22 21 54" src="https://user-images.githubusercontent.com/96298768/147006658-3fae14fd-0533-4c10-a970-42cf0b4d4b22.png">
<img width="294" alt="Screenshot 2021-12-21 at 22 21 40" src="https://user-images.githubusercontent.com/96298768/147006789-d21ea3bd-9da3-40d4-bba2-bf7993c1aa85.png">

### **So in summary**

- Civetweb 1.16 (master) - I've tried 1.15
- Open SSL 1.1.1k
- Mac OSX
- JMeter for saturation of the server
- High volume of WebSocket messages
- SSL Activated 
- Open SSL Statically compiled
- I can't get the bundled example WebSocket server to fall over even after allowing the count to reach nearly a million.

Stumped. Any help would be appreciated.

**Comment by t-j-h** [comment]
Do an ASAN build or use valgrind.

**Comment by jackfruit2** [comment]
Thanks @t-j-h , will have a go with valgrind.

**Comment by jackfruit2** [comment]
@t-j-h - I'm fairly new to C++ (6 months or so) and your advice helped me massively smoke out the issue. It was a bug in my test harness, stomping on a bit of memory I shouldn't be. Knowledge of the Address Sanitiser is best xmas present for a newbie C++ developer.  Merry xmas and happy new year!!

**Comment by t-j-h** [comment]
@jackfruit2 you're welcome - developing C or C++ code without things like ASAN, valgrind (and things like purify and coverity prevent as commercial tools) is doing things the hard way ... enjoy writing code with tools that help you find your errors.

Generally with well curated open source projects, typically this sort of error is in the users code (not always - just typically) especially then the project has a test suite and wide usage.
